notifyme-sdk-backend/notifyme-sdk-backend-ws/src/main/java/ch/ubique/notifyme/sdk/backend/ws/util/CryptoUtilV3.java

```diff
@@ -0,0 +1,320 @@
+            return new NoncesAndNotificationKey(noncePreId, nonceTimekey, notificationKey);
+        } catch (GeneralSecurityException e) {
+            throw new RuntimeException("HKDF threw GeneralSecurityException");
+        }
+    }
+
+    public IBECiphertext encryptInternal(G2 masterPublicKey, byte[] identity, byte[] message) {
+
+        byte[] x = createNonce(NONCE_BYTES);
+
+        Fr r = new Fr();
+        r.setHashOf(concatenate(x, identity, message));
+
+        G2 c1 = new G2();
+        Mcl.mul(c1, baseG2(), r);
+
+        G1 g1_temp = new G1();
+        Mcl.hashAndMapToG1(g1_temp, identity);
+
+        GT gt1_temp = new GT();
+        Mcl.pairing(gt1_temp, g1_temp, masterPublicKey);
+
+        GT gt_temp = new GT();
+        Mcl.pow(gt_temp, gt1_temp, r);
+
+        byte[] c2_pair = cryptoHashSHA256(gt_temp.serialize());
+        byte[] c2 = xor(x, c2_pair);
+
+        byte[] nonce = createNonce();
+
+        byte[] c3 = cryptoSecretboxEasy(cryptoHashSHA256(x), message, nonce);
+
+        return new IBECiphertext(c1.serialize(), c2, c3, nonce);
+    }
+
+    public byte[] decryptInternal(
+            IBECiphertext ibeCiphertext, G1 secretKeyForIdentity, byte[] identity) {
+        G2 c1 = new G2();
+        c1.deserialize(ibeCiphertext.getC1());
+
+        GT gt_temp = new GT();
+        Mcl.pairing(gt_temp, secretKeyForIdentity, c1);
+
+        byte[] hash = cryptoHashSHA256(gt_temp.serialize());
+        byte[] x_p = xor(ibeCiphertext.getC2(), hash);
+
+        byte[] msg_p = new byte[ibeCiphertext.getC3().length - Box.MACBYTES];
+        int result =
+                sodium.crypto_secretbox_open_easy(
+                        msg_p,
+                        ibeCiphertext.getC3(),
+                        ibeCiphertext.getC3().length,
+                        ibeCiphertext.getNonce(),
+                        cryptoHashSHA256(x_p));
+        if (result != 0) return null;
+
+        // Additional verification
+        Fr r_p = new Fr();
+        r_p.setHashOf(concatenate(x_p, identity, msg_p));
+
+        G2 c1_p = new G2();
+        Mcl.mul(c1_p, baseG2(), r_p);
+
+        if (!c1.equals(c1_p)) {
+            return null;
+        }
+
+        if (!secretKeyForIdentity.isValidOrder() || secretKeyForIdentity.isZero()) {
+            return null;
+        }
+        return msg_p;
+    }
+
+    public byte[] cryptoSecretboxOpenEasy(byte[] key, byte[] cipherText, byte[] nonce) {
+        final var decryptedMessage = new byte[cipherText.length - Box.MACBYTES];
+        int result =
+                sodium.crypto_secretbox_open_easy(
+                        decryptedMessage, cipherText, cipherText.length, nonce, key);
+        if (result != 0) return new byte[0];
+        return decryptedMessage;
+    }
+
+    public byte[] encryptAssociatedData(
+            byte[] secretKey,
+            String message,
+            byte[] countryData,
+            byte[] nonce,
```

Comment:
the order of arguments should be changed, so that it is secretKey, nonce and then all the payload arguments, otherwise it is very confusing why we need the nonce here.

Reply:
Resolved: See [https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/ea7769558a7b37b814e285e00c608b17822bbaab](https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/ea7769558a7b37b814e285e00c608b17822bbaab)